alphacsc/_encoder.py

```diff
@@ -0,0 +1,336 @@
+    n_atoms : int
+        The number of atoms to learn.
+    atom_support : int
+        The support of the atom.
+    algorithm : 'batch' | 'greedy' | 'online' | 'stochastic'
+        Dictionary learning algorithm.
+    reg : float
+        The regularization parameter.
+    loss : {{ 'l2' | 'dtw' | 'whitening'}}
+        Loss for the data-fit term. Either the norm l2 or the soft-DTW.
+    loss_params : dict
+        Parameters of the loss.
+    uv_constraint : {{'joint' | 'separate'}}
+        The kind of norm constraint on the atoms:
+
+        - :code:`'joint'`: the constraint is ||[u, v]||_2 <= 1
+        - :code:`'separate'`: the constraint is ||u||_2 <= 1 and ||v||_2 <= 1
+    feasible_evaluation : boolean
+        If feasible_evaluation is True, it first projects on the feasible set,
+        i.e. norm(uv_hat) <= 1.
+    n_jobs : int
+        The number of parallel jobs.
+    use_sparse_z : boolean
+        Use sparse lil_matrices to store the activations.
+
+    Returns
+    -------
+    enc - a ZEncoder instance
+
+    Example usage
+    -------------
+    with get_encoder_for('lgcd') as enc:
+        ...
+    """
+    assert isinstance(z_kwargs, dict), 'z_kwargs should be a valid dictionary.'
+
+    assert (X is not None and len(X.shape) == 3), \
+        'X should be a valid array of shape (n_trials, n_channels, n_times).'
+
+    assert (D_hat is not None and len(D_hat.shape) in [2, 3]), \
+        'D_hat should be a valid array of shape ' \
+        '(n_trials, n_channels, n_times) ' \
+        'or (n_atoms, n_channels + atom_support).'
+
+    assert algorithm in ['batch', 'greedy', 'online', 'stochastic'], \
+        f'unrecognized algorithm type: {algorithm}.'
+
+    # ASK can reg be less than 0?
+    assert reg is not None, 'reg value cannot be None.'
+
+    assert loss in ['l2', 'dtw', 'whitening'], \
+        f'unrecognized loss type: {loss}.'
+
+    assert isinstance(
+        loss_params, dict), 'loss_params should be a valid dictionary.'
+
+    assert uv_constraint in ['joint', 'separate'], \
+        f'unrecognized uv_constraint type: {uv_constraint}.'
+    # ASK check invalid values for n_atoms, atom_support, n_jobs
+
+    if solver in ['l-bfgs', 'lgcd']:
+        return AlphaCSCEncoder(
+            solver,
+            z_kwargs,
+            X,
+            D_hat,
+            n_atoms,
+            atom_support,
+            algorithm,
+            reg,
+            loss,
+            loss_params,
+            uv_constraint,
+            feasible_evaluation,
+            n_jobs,
+            use_sparse_z)
+    else:
+        raise ValueError(f'unrecognized solver type: {solver}.')
+
+
+class BaseZEncoder:
+    def compute_z(self):
+        """
+        Perform one incremental z update.
+        This is the "main" function of the algorithm.
+        """
+        raise NotImplementedError()
+
+    def compute_z_partial(self, i0):
+        """
+        Compute z on a slice of the signal X,
+        for online learning.
```

Comment:
Doesn't this work on one line?
```suggestion
        Compute z on a slice of the signal X, for online learning.
```